### ISSUE TITLE
Dependencies defined in initialize method doesn't work as expected

### ISSUE THREAD

**Issue by jfrac** [collectiveidea/delayed_job#1136]
Hello, I posted the same issue in the `google-drive-ruby` repository about that topic, but I think it could be more related with delayed jobs. I found 2 problems so far:

1. RuntimeError: THIS IS NOT A RSA! https://github.com/gimite/google-drive-ruby/issues/389

```ruby
# frozen_string_literal: true

class GoogleSpreadsheetClient
  SCOPE = ["https://www.googleapis.com/auth/drive", "https://spreadsheets.google.com/feeds/"].freeze
  def initialize
    credentials = Google::Auth.get_application_default(SCOPE)
    @client ||= GoogleDrive::Session.new(credentials, nil, nil, nil)
  end

  def invoke(spreadsheet_title, worksheet_title, object)
    worksheet = get_worksheet(spreadsheet_title, worksheet_title)
    worksheet.update_cells(worksheet.num_rows + 1, 1, [object.to_row])
    worksheet.save
  end

  def get_worksheet(spreadsheet_title, worksheet_title)
    spreadsheet = @client.spreadsheet_by_title(title)
    spreadsheet.worksheet_by_title(worksheet_title)
  end
end
```

2. Defining `@logger = Rails.logger` inside `initialize` provoques the logger don't log anything when I use it in a delayed method like `@logger.info(message)` 

```ruby
# frozen_string_literal: true

class Service
  def initialize
    @logger = Rails.logger
  end

  def invoke
    ...
      @logger.info("message") # it doesn't log anything
    ...
  end
end
```

Basically, when I define an object instance inside de `initialize` method, and then run the delayed method, I get some errors or an unexpected behaviour. I'm not sure if defining instances in a delayed object is supported by `delayed_job` gem.

**Comment by albus522** [comment]
If you are calling delay on an instance object of Service, I suspect the Logger instance is not serializing/deserializing well. With any background job processing, you should always strive to serialize the simplest object possible as you will eventually run into issues. Just because DJ can usually run delay on more complex data doesn't mean you should.

In the above example, why are you doing `@logger = Rails.logger` vs using `Rails.logger` directly where you need it?

**Comment by jfrac** [comment]
> In the above example, why are you doing @logger = Rails.logger vs using Rails.logger directly where you need it?

In that case, just to have all the dependencies in the constructor.

**Comment by albus522** [comment]
Closing stale
